faiss_vector_cache.go

```diff
@@ -52,34 +52,56 @@ func (vc *vectorIndexCache) Clear() {
 	vc.m.Unlock()
 }
 
-func (vc *vectorIndexCache) loadOrCreate(fieldID uint16, mem []byte, except *roaring.Bitmap) (
-	index *faiss.IndexImpl, vecDocIDMap map[int64]uint32, vecIDsToExclude []int64, err error) {
+func (vc *vectorIndexCache) loadOrCreate(fieldID uint16, mem []byte,
+	loadDocVecIDMap bool, except *roaring.Bitmap) (
+	index *faiss.IndexImpl, vecDocIDMap map[int64]uint32, docVecIDMap map[uint32]int64,
+	vecIDsToExclude []int64, err error) {
 	var found bool
-	index, vecDocIDMap, vecIDsToExclude, found = vc.loadFromCache(fieldID, except)
+	index, vecDocIDMap, docVecIDMap, vecIDsToExclude, found = vc.loadFromCache(
+		fieldID, except)
```

Comment:
Hmm, something doesn't feel right here - I think it is the situation if the first loadFromCache didn't need us to "loadDocVecIDMap" and while this is still in cache the 2nd invocation requires it .. in which case shouldn't you need to invalidate the cache and rebuild it?

Reply:
good catch, thanks!

Reply:
Thinking it make be a good idea to integrate loadFromCache and createAndCache into one method - loadFromCache which will also be responsible for creation if it's a cache miss. This would avoid us having to check `!found || (loadDocVecIDMap && docVecIDMap == nil)` twice otherwise.

Do you see any obstacles with this approach?

Reply:
none as yet, sounds good.

Reply:
I think this approach still has the same downside as the earlier approach - the first pre-filtering query will require rebuilding everything other than the docVecIDMap.
I'll try to refactor so that the first pre-filtered query only constructs the docVecIDMap using the vecDocIDMap.

Reply:
refactored this part, pls take a look at recent commits.